packages/iov-keycontrol/src/connection.ts

```diff
@@ -0,0 +1,232 @@
+class SendProducer<T> implements Producer<T> {
+  // tslint:disable-next-line:readonly-keyword
+  private listener: Listener<T> | undefined;
+
+  public start(listener: Listener<T>): void {
+    // tslint:disable-next-line:no-object-mutation
+    this.listener = listener;
+  }
+
+  public stop(): void {
+    if (this.listener) {
+      this.listener.complete();
+      // tslint:disable-next-line:no-object-mutation
+      this.listener = undefined;
+    }
+  }
+
+  public send(msg: T): void {
+    if (this.listener) {
+      this.listener.next(msg);
+    }
+  }
+}
+
+type EventProducer = SendProducer<Event>;
+type MessageProducer = SendProducer<Message>;
+
+// localConnectionPair returns a pair of connections that can be
+// pass to Client and Server for them to communicate.
+//
+// Generally thought of for tests, but could also be used
+// to isolate components within one process.
+export const localConnectionPair = (): [Connection, Connection] => {
+  const a = new SendProducer<Message>();
+  const b = new SendProducer<Message>();
+  return [localConnection(a, b), localConnection(b, a)];
+};
+
+const localConnection = (client: MessageProducer, server: MessageProducer): Connection => ({
+  send: client.send,
+  receive: Stream.create(server),
+  disconnect: () => client.stop(),
+});
+
+// Promiser holds the callbacks to later resolve Promise from different
+// control flow
+interface Promiser {
+  readonly resolve: (value: any) => void;
+  readonly reject: (err: any) => void;
+}
+
+// Client sends requests to a connection and gets responses,
+// possibly out of order
+export class Client {
+  private readonly connection: Connection;
+  private readonly resolvers: Map<string, Promiser>;
+  private readonly streams: Map<string, EventProducer>;
+
+  constructor(connection: Connection) {
+    this.connection = connection;
+    this.resolvers = new Map<string, Promiser>();
+    this.streams = new Map<string, EventProducer>();
+    this.listen();
+  }
+
+  public request(method: string, params: any): Promise<any> {
+    const req: RequestMessage = {
+      ...envelope(),
+      kind: MessageKind.REQUEST,
+      method,
+      params,
+    };
+    this.connection.send(req);
+    return new Promise<any>((resolve, reject) => {
+      // make sure nothing already registered
+      if (this.resolvers.get(req.id)) {
+        reject(`Cannot reuse ${req.id}, request currently pending`);
+      } else {
+        this.resolvers.set(req.id, { resolve, reject });
+      }
+    });
+  }
+
+  public subscribe(query: string): Stream<Event> {
+    const subscriptionId = randomId();
+
+    const sub: SubscribeMessage = {
+      ...envelope(),
+      kind: MessageKind.REQUEST,
+      method: "subscribe",
+      params: {
+        query,
+        subscriptionId,
+      },
+    };
+    this.connection.send(sub);
+    const producer = new SendProducer<Event>();
+    // TODO: somehow store this to unsubscribe if subscription request
+    // above returns an error
```

Comment:
Or make subscribe async, returning `Promise<Stream<Event>>`. This guarantees a `Stream<Event>` in case of success and a rejection in case of error. No complicated error handling in different places.

Reply:
`Promise<Stream<Event>>` seems like a beast, and I bet anyone using pure js will be bit by it, but it does make sense. The promise will reject on connection error (bad query), the stream on later error. I was wondering about using the stream error to signal bad query string.

Thanks for the suggestion.